### PR TITLE
User Subcomposables

### LINKED ISSUE #4809: Implement factories for user subcomposables
## What is the motivation for adding / enhancing this feature?

Add missing interfaces for user subcomposales
```ts
export interface UseUserShipping<ADDRESS> {
  addresses: ComputedProperty<ADDRESS[]>;
  totalAddresses: ComputedProperty<number>;
  addAddress: (address: ADDRESS) => Promise<void>;
  deleteAddress: (address: ADDRESS) => Promise<void>;
  updateAddress: (address: ADDRESS) => Promise<void>;
  load: () => Promise<void>;
  defaultAddress: ComputedProperty<ADDRESS>;
  setDefault (address: ADDRESS);
  loading: ComputedProperty<boolean>;
}
```

```ts
export interface UseUserBilling<ADDRESS> {
  addresses: ComputedProperty<ADDRESS[]>;
  totalAddresses: ComputedProperty<number>;
  addAddress: (address: ADDRESS) => Promise<void>;
  deleteAddress: (address: ADDRESS) => Promise<void>;
  updateAddress: (address: ADDRESS) => Promise<void>;
  load: () => Promise<void>;
  defaultAddress: ComputedProperty<ADDRESS>;
  setDefault (address: ADDRESS);
  loading: ComputedProperty<boolean>;
}
```

## What are the acceptance criteria 

- [x] create typings in `core` `
- [x] create factories
- [x] cover them with unit tests
- [x] contemplate great unit tests coverage 😎 

## Version of Vue Storefront

- [ ] Vue Storefront
- [x] Vue Storefront Next

## Can you complete this feature request by yourself?

- [ ] YES
- [x] NO

### PR DESCRIPTION
**Pull request by Fifciu**
- [x] Shipping Composable
- [x] Billing Composable
- [x] tests

Close #4809

### CODE REVIEW

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import { Ref, computed } from '@vue/composition-api';
+import { UseUserBilling } from '../types';
+import { sharedRef } from '../utils';
+
+export interface UseUserBillingFactoryParams<ADDRESS> {
+  addAddress: (address: ADDRESS) => Promise<ADDRESS[]>;
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:6]: Please use single param here (like in other factories)
```
addAddress({ params })
```
this way if we add/reorder something its not a breaking change

Also you should provide access to current state here (`addressess`) and mark it as readonly.Give people everything they could need in this function from current composable.

Above applies to whole `factoryParams` object

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  deleteAddress: (address: ADDRESS) => Promise<ADDRESS[]>;
+  updateAddress: (address: ADDRESS) => Promise<ADDRESS[]>;
+  load: () => Promise<ADDRESS[]>;
+  setDefault: (address: ADDRESS) => Promise<void>;
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:10]: you should return the new default address so in the factory its picked up and assigned to a variable

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+interface UseUserShippingFactory<ADDRESS> {
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:13]: image

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      } finally {
+        loading.value = false;
+      }
+    };
+
+    const setDefault = async (address: ADDRESS) => {
+      loading.value = true;
+      try {
+        await factoryParams.setDefault(address);
+        defaultAddress.value = address;
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:74]: always assign to value returned by `factoryParams`. If integrator decides to strip down some properties or do some transformations they will be omitted

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 1 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useUserBillingFactory.ts:6]: Ok, did it like that:
```ts
const readonlyAddresses = computed(() => addresses.value);
// ...
addresses.value = await factoryParams.addAddress({
          address,
          addresses: readonlyAddresses
        });
```

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
+import { Ref, ComputedRef, computed } from '@vue/composition-api';
+import { UseUserBilling } from '../types';
+import { sharedRef } from '../utils';
+
+export interface UseUserBillingFactoryParams<ADDRESS> {
+  addAddress: (params: {
+    address: ADDRESS;
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:7]: mark everything here as readonly

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    addresses: ComputedRef<ADDRESS[]>;
+  }) => Promise<ADDRESS[]>;
+  updateAddress: (params: {
+    address: ADDRESS;
+    addresses: ComputedRef<ADDRESS[]>;
+  }) => Promise<ADDRESS[]>;
+  load: (params: {
+    addresses: ComputedRef<ADDRESS[]>;
+  }) => Promise<ADDRESS[]>;
+  setDefault: (params: {
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:21]: we will need current state here as well (so value of `defaultAddress`)

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 5 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useUserBillingFactory.ts:7]: Marked as readonly from TS

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 6 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useUserBillingFactory.ts:23]: Also added to updateAddress and deleteAddress based on our conversation

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+export const useUserBillingFactory = <ADDRESS>(
+  factoryParams: UseUserBillingFactoryParams<ADDRESS>
+) => {
+
+  const useUserBilling = (): UseUserBilling<ADDRESS> => {
+    const defaultAddress: Ref<ADDRESS> = sharedRef(null, 'useUserBilling-default-address');
+    const loading: Ref<boolean> = sharedRef(false, 'useUserBilling-loading');
+    const addresses: Ref<ADDRESS[]> = sharedRef([], 'useUserBilling-addresses');
+
+    const readonlyAddresses: ComputedRef<ADDRESS[]> = computed(() => addresses.value);
```
Comment (filrak) [inline packages/core/core/src/factories/useUserBillingFactory.ts:39]: using typescript `readonly` in functions seems more sufficient for this job

**Review by andrzejewsky — APPROVED** [review]

File: packages/core/core/src/factories/useUserBillingFactory.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      }
+    };
+
+    const setDefault = async (address: ADDRESS) => {
+      loading.value = true;
+      try {
+        defaultAddress.value = await factoryParams.setDefault({
+          address,
+          defaultAddress: readonlyDefaultAddress,
+          addresses: readonlyAddresses
```
Comment (andrzejewsky) [inline packages/core/core/src/factories/useUserBillingFactory.ts:105]: please unwrap variables

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 8 above)
Comment (andrzejewsky) [reply packages/core/core/src/factories/useUserBillingFactory.ts:105]: everywhere you call factory params

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 7 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useUserBillingFactory.ts]:
```ts
const readonlyDefaultAddress: Readonly<ADDRESS> = unref(defaultAddress);
```

File: packages/core/core/src/factories/useUserBillingFactory.ts (hunk 8 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useUserBillingFactory.ts:105]:
```ts
const readonlyDefaultAddress: Readonly<ADDRESS> = unref(defaultAddress);
```

**Review by andrzejewsky — APPROVED** [review]

**Review by filrak — APPROVED** [review]

**Review by filrak — APPROVED** [review]

**Pull request merged**